homeassistant/components/tami4/sensor.py

```diff
@@ -0,0 +1,157 @@
+        key="uv_last_replacement",
+        name="UV Last Replacement",
+        icon="mdi:calendar",
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.DATE,
+    ),
+    SensorEntityDescription(
+        key="uv_upcoming_replacement",
+        name="UV Upcoming Replacement",
+        icon="mdi:calendar",
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.DATE,
+    ),
+    SensorEntityDescription(
+        key="uv_status",
+        name="UV Status",
+        icon="mdi:clipboard-check-multiple",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key="filter_last_replacement",
+        name="Filter Last Replacement",
+        icon="mdi:calendar",
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.DATE,
+    ),
+    SensorEntityDescription(
+        key="filter_upcoming_replacement",
+        name="Filter Upcoming Replacement",
+        icon="mdi:calendar",
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.DATE,
+    ),
+    SensorEntityDescription(
+        key="filter_status",
+        name="Filter Status",
+        icon="mdi:clipboard-check-multiple",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key="filter_milli_litters_passed",
+        name="Filter Water Passed",
+        native_unit_of_measurement=VOLUME_MILLILITERS,
+        icon="mdi:water",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+]
+
+
+class Tami4EdgeWaterQualityCoordinator(DataUpdateCoordinator):
+    """Tami4Edge water quality coordinator."""
+
+    def __init__(self, hass: HomeAssistant, edge: Tami4EdgeAPI) -> None:
+        """Initialize the water quality coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Tami4Edge water quality coordinator",
+            update_interval=timedelta(minutes=10),
+        )
+        self.edge = edge
+
+    async def _async_update_data(self) -> dict:
+        """Fetch data from the API endpoint."""
+        try:
+            water_quality = await self.hass.async_add_executor_job(
+                self.edge.get_water_quality
+            )
+            return {
+                "uv_last_replacement": water_quality.uv.last_replacement,
+                "uv_upcoming_replacement": water_quality.uv.upcoming_replacement,
+                "uv_status": water_quality.uv.status,
+                "filter_last_replacement": water_quality.filter.last_replacement,
+                "filter_upcoming_replacement": water_quality.filter.upcoming_replacement,
+                "filter_status": water_quality.filter.status,
+                "filter_milli_litters_passed": water_quality.filter.milli_litters_passed,
+            }
+        except Exception as ex:
+            raise UpdateFailed(f"Error communicating with API: {ex}") from ex
```

Comment:
We generally put the coordinator in its own module. Please consider moving this to a `coordinator.py` file.